iot-hub/Quickstarts/read-d2c-messages/ReadDeviceToCloudMessages.js

```diff
@@ -1,55 +1,159 @@
+    port: 5671,
+    reconnect: false,
+    password: token,
+  };
+
+  const connection = new Connection(connectionOptions);
+  await connection.open();
+
+  // Create the receiver that will trigger a redirect error.
+  const receiver = await connection.createReceiver({
+    source: { address: `amqps://${HostName}/messages/events/$management` },
+  });
+
+  return new Promise((resolve, reject) => {
+    receiver.on(ReceiverEvents.receiverError, (context) => {
+      const error = context.receiver && context.receiver.error;
+      if (isAmqpError(error) && error.condition === "amqp:link:redirect") {
+        const hostname = error.info && error.info.hostname;
+        if (!hostname) {
+          reject(error);
+        } else {
+          resolve(
+            `Endpoint=sb://${hostname}/;EntityPath=${iotHubName};SharedAccessKeyName=${SharedAccessKeyName};SharedAccessKey=${SharedAccessKey}`
+          );
+        }
+      } else {
+        reject(error);
+      }
+      connection.close().catch(() => {
+        /* ignore error */
+      });
+    });
+  });
+}
 
 var printError = function (err) {
   console.log(err.message);
 };
 
 // Display the message content - telemetry and properties.
 // - Telemetry is sent in the message body
-// - The device can add arbitrary application properties to the message
+// - The device can add arbitrary properties to the message
 // - IoT Hub adds system properties, such as Device Id, to the message.
 var printMessage = function (message) {
-  console.log('Telemetry received: ');
+  console.log("Telemetry received: ");
   console.log(JSON.stringify(message.body));
-  console.log('Application properties (set by device): ')
-  console.log(JSON.stringify(message.applicationProperties));
-  console.log('System properties (set by IoT Hub): ')
-  console.log(JSON.stringify(message.annotations));
-  console.log('');
+  console.log("Properties (set by device): ");
+  console.log(JSON.stringify(message.userProperties));
+  console.log("System properties (set by IoT Hub): ");
+  console.log(JSON.stringify(message.systemProperties));
+  console.log("");
 };
 
-// Connect to the partitions on the IoT Hub's Event Hubs-compatible endpoint.
-// This example only reads messages sent after this application started.
-var ehClient;
-EventHubClient.createFromIotHubConnectionString(connectionString).then(function (client) {
-  console.log("Successfully created the EventHub Client from iothub connection string.");
-  ehClient = client;
-  return ehClient.getPartitionIds();
-}).then(function (ids) {
-  console.log("The partition ids are: ", ids);
-  return ids.map(function (id) {
-    return ehClient.receive(id, printMessage, printError, { eventPosition: EventPosition.fromEnqueuedTime(Date.now()) });
+async function main() {
+  console.log("IoT Hub Quickstarts - Read device to cloud messages.");
+
+  const iotHubConnectionString = "{your Iot Hub connection string}";
+  const iotHubName = "{your Iot Hub name}";
```

Comment:
iotHubName isn't needed because both the built-in endpoint and the endpoint returned from an IotHub connection string include the `EntityPath`.

Reply:
I chose to keep this for the case when user does have access to the Event Hubs compatible connection string and it does not contain the EntityPath. The portal does have the EntityPath included, am not sure if the Azure CLI does as well...

Reply:
Turns out you have to use the Azure CLI thrice, once for the endpoint, once for the entity path (which is the same as the iothub name for now), and one more time to get the sharedaccesskey :(

Reply:
https://github.com/Azure-Samples/azure-iot-samples-node/pull/38/commits/31a4a4452849c2fc10165ece1b3489bea7e3655b should fix it

Added a helper for forming connection string using the results from Azure CLI. This will ensure that `EntityPath` is part of the connection string and we wont need to pass the path around